cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/EOGWorklist.kt

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright (c) 2023, Fraunhofer AISEC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.helpers
+
+import de.fraunhofer.aisec.cpg.graph.Node
+import de.fraunhofer.aisec.cpg.graph.edge.PropertyEdge
+import java.util.IdentityHashMap
+
+/**
+ * A complete lattice is an ordered structure of values of type [T]. [T] could be anything, e.g., a
+ * set, a new data structure (like a range), or anything else. [T] depends on the analysis and
+ * typically has to abstract the value for the specific purpose.
+ *
+ * This class is actually used to hold individual instances of the lattice's elements and to compute
+ * bigger elements depending on these two elements.
+ *
+ * Implementations of this class have to implement the comparator, the least upper bound of two
+ * lattices.
+ */
+abstract class Lattice<T>(open val elements: T) : Comparable<Lattice<T>?> {
+    /**
+     * Computes the least upper bound of this lattice and [other]. It returns a new object and does
+     * not modify either of the objects.
+     */
+    abstract fun lub(other: Lattice<T>?): Lattice<T>
+
+    /** Duplicates the object, i.e., makes a deep copy. */
+    abstract fun duplicate(): Lattice<T>
+}
+
+/**
+ * Implements the [Lattice] over a set of nodes. The lattice itself is constructed by the powerset.
+ */
+class PowersetLattice(override val elements: Set<Node>) : Lattice<Set<Node>>(elements) {
+    override fun lub(other: Lattice<Set<Node>>?) =
+        PowersetLattice((other?.elements ?: setOf()).union(this.elements))
+
+    override fun duplicate() = PowersetLattice(this.elements.toSet())
+
+    override fun compareTo(other: Lattice<Set<Node>>?): Int {
```

Comment:
compareTo confuses me. Is this a workaround in order to compute the greatest lower bound? because for the powersets glb we could just use `intersect` if I understand lattices correctly 🤯

Reply:
It's used to determine if we have to continue with the iteration, update something, etc. (e.g., if the new element > current element, we have to continue). By definition, a group has an ordering operator, which we model it with the `compareTo`.